core/src/main/scala/pureconfig/ConfigSource.scala

```
@@ -0,0 +1,87 @@
+package pureconfig
+
+import java.io.File
+import java.net.URL
+import java.nio.file.Path
+
+import scala.reflect.ClassTag
+
+import com.typesafe.config._
+import pureconfig.ConfigReader.Result
+import pureconfig.backend.{ ConfigFactoryWrapper, PathUtil }
+import pureconfig.error.{ ConfigReaderException, ConfigReaderFailures }
+
+trait ConfigSource { outer =>
+  def value: Result[ConfigValue]
+
+  def cursor: Result[ConfigCursor] =
+    value.right.map(ConfigCursor(_, Nil))
+
+  def fluentCursor: FluentConfigCursor =
+    FluentConfigCursor(cursor)
+
+  def at(namespace: String): ConfigSource =
+    ConfigSource.fromCursor(fluentCursor.at(PathUtil.splitPath(namespace).map(p => p: PathSegment): _*))
+
+  def load[A](implicit reader: Derivation[ConfigReader[A]]): Result[A] =
+    cursor.right.flatMap(reader.value.from)
+
+  @throws[ConfigReaderException[_]]
+  def loadOrThrow[A: ClassTag](implicit reader: Derivation[ConfigReader[A]]): A = {
+    load[A] match {
+      case Right(config) => config
+      case Left(failures) => throw new ConfigReaderException[A](failures)
+    }
+  }
+}
+
+class ConfigObjectSource(readConfig: => Result[Config]) extends ConfigSource {
+  private lazy val config: Result[Config] = readConfig
+  lazy val value: Result[ConfigObject] = config.right.map(_.resolve.root)
+
+  def withFallback(cs: ConfigObjectSource): ConfigObjectSource =
+    new ConfigObjectSource(Result.zipWith(config, cs.config)(_.withFallback(_)))
+
+  def withOptionalFallback(cs: ConfigObjectSource): ConfigObjectSource =
+    withFallback(cs.recoverWith({ case _ => Right(ConfigFactory.empty) }))
```

This is not very intuitive. I think that if the config file is invalid (rather than missing), most of the users would want to report the errors. So we should handle those two conditions separately.

Also, I think we should factor out a more fundamental primitive:
```scala
def optional: ConfigObjectSource =
  cs.recoverWith({ case _: CannotRead => Right(ConfigFactory.empty) })

def withOptionalFallback(cs: ConfigObjectSource): ConfigObjectSource =
  withFallback(cs.optional)
```

---

Fair enough. I'll remove `withOptionalFallback` altogether, as it doesn't seem to provide a lot of benefits if `optional` exists.